src/sources/mod.rs

```diff
@@ -356,15 +379,57 @@ pub(crate) trait EventDispatcher<Data> {
         data: &mut Data,
     ) -> crate::Result<PostAction>;
 
-    fn register(&self, poll: &mut Poll, token_factory: &mut TokenFactory) -> crate::Result<()>;
+    fn register(
+        &self,
+        poll: &mut Poll,
+        additional_lifetime_register: AdditionalLifetimeEventsRegister<'_>,
+        token_factory: &mut TokenFactory,
+    ) -> crate::Result<()>;
 
-    fn reregister(&self, poll: &mut Poll, token_factory: &mut TokenFactory) -> crate::Result<bool>;
+    fn reregister(
+        &self,
+        poll: &mut Poll,
+        additional_lifetime_register: AdditionalLifetimeEventsRegister<'_>,
+        token_factory: &mut TokenFactory,
+    ) -> crate::Result<bool>;
+
+    fn unregister(
+        &self,
+        poll: &mut Poll,
+        additional_lifetime_register: AdditionalLifetimeEventsRegister<'_>,
+    ) -> crate::Result<bool>;
+
+    fn before_sleep(&self) -> crate::Result<Option<(Readiness, Token)>>;
+    fn before_handle_events(&self, was_awoken: bool);
+}
+
+#[derive(Default)]
+pub(crate) struct AdditionalLifetimeEventsSet {
+    pub(crate) values: RefCell<Vec<(RegistrationToken, bool)>>,
+}
+
+impl AdditionalLifetimeEventsSet {
+    pub(crate) fn create_register_for_token(
+        &self,
+        token: RegistrationToken,
+    ) -> AdditionalLifetimeEventsRegister {
+        AdditionalLifetimeEventsRegister { set: self, token }
+    }
+}
 
-    fn unregister(&self, poll: &mut Poll) -> crate::Result<bool>;
+pub(crate) struct AdditionalLifetimeEventsRegister<'a> {
+    set: &'a AdditionalLifetimeEventsSet,
+    token: RegistrationToken,
+}
 
-    fn pre_run(&self, data: &mut Data) -> crate::Result<()>;
+impl<'a> AdditionalLifetimeEventsRegister<'a> {
+    fn register(&mut self) {
+        self.set.values.borrow_mut().push((self.token, false))
+    }
 
-    fn post_run(&self, data: &mut Data) -> crate::Result<()>;
+    fn unregister(&mut self) {
+        self.set.values.borrow_mut().retain(|it| it.0 != self.token)
+    }
 }
```

Comment:
Could add a few comments over these types and methods, to keep a trace of what they are for and how they are used? That'd make future maintenance of this code easier.